### PR TITLE
Zscaler ZIA enhancement: Add IP Destination group commands

### PR DESCRIPTION
**Pull request by rootbid**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Add support for IP Destination groups management for Zscaler ZIA integration.

## Screenshots
MicrosoftTeams-image (5)
MicrosoftTeams-image (4)
MicrosoftTeams-image (3)
MicrosoftTeams-image (2)
MicrosoftTeams-image (1)

## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by pallavi-garad** [comment]
These are the updated screenshots for delete-ip-destination-groups and list-ip-destination-groups with lite flag.

image
image
image
image

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 1]
```diff
@@ -1016,13 +1017,298 @@ def set_user_command(args):
...
+        'ContentsFormat': formats['json'],
+        'ReadableContentsFormat': formats['markdown'],
+        'HumanReadable': tableToMarkdown("IPv4 Destination group created",
+                                         content, headers, removeNull=True),
+        'EntryContext': {'Zscaler.IPDestinationGroup': content},
+    }
+    return entry
+
+
+def list_ip_destination_groups(args: dict):
```
Comment (dantavori) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.py:1059]: there's a lot of repetative code in this method, we can do something like:
```
lite_endpoint = "/lite" if lite else ""
ipv4_cmd_url = "/ipDestinationGroups" + lite_endpoint
ipv4_responses = http_request('GET', ipv4_cmd_url).json()
ipv4_contents = get_contents_lite(ipv4_responses) if lite else get_contents(ipv4_responses)
...
if include_ipv6:
    ipv6_cmd_url = "/ipDestinationGroups/ipv6DestinationGroups" + lite_endpoint
    ipv6_responses = http_request('GET', ipv6_cmd_url).json()
    ipv6_contents = get_contents_lite(ipv6_responses) if lite else get_contents(ipv6_responses)
    ...
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 2]
```diff
@@ -61,15 +61,16 @@
...
 )
 
 ''' HANDLE PROXY '''
-if not PROXY:  # pragma: no cover
-    del os.environ['HTTP_PROXY']
-    del os.environ['HTTPS_PROXY']
-    del os.environ['http_proxy']
-    del os.environ['https_proxy']
+# Remove proxy if not set to true in params
+handle_proxy()
```
Comment (dantavori) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.py:73]: nice!

**Review by dantavori — CHANGES_REQUESTED** [review]
looks good, thank you for your contribution!
see my notes - some of them are relevant to more than one command.

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py (hunk 1 above)
Comment (rootbid) [reply Packs/Zscaler/Integrations/Zscaler/Zscaler.py:1059]: done

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 3]
```diff
@@ -1016,13 +1017,298 @@ def set_user_command(args):
...
+    entry = {
+        'Type': entryTypes['note'],
+        'Contents': content,
+        'ContentsFormat': formats['json'],
+        'ReadableContentsFormat': formats['markdown'],
+        'HumanReadable': tableToMarkdown("IPv4 Destination group created",
+                                         content, headers, removeNull=True),
+        'EntryContext': {'Zscaler.IPDestinationGroup': content},
+    }
+    return entry
```
[the remark this answers is not in the document]
Comment (rootbid) [reply Packs/Zscaler/Integrations/Zscaler/Zscaler.py]: done

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 4]
```diff
@@ -1016,13 +1017,298 @@ def set_user_command(args):
...
+        'description': args.get('description', ''),
+        'addresses': argToList(args.get('addresses', '')),
+        'isNonEditable': args.get('is_non_editable', False)
+    }
+    cmd_url = "/ipDestinationGroups"
+    try:
+        response = http_request('POST', cmd_url, data=json.dumps(payload),
+                                headers=DEFAULT_HEADERS).json()
+    except json.decoder.JSONDecodeError as exp:
+        return_error(f'Failed to execute zscaler-create-ip-destination-group command. Error: {str(exp)}')
```
[the remark this answers is not in the document]
Comment (rootbid) [reply Packs/Zscaler/Integrations/Zscaler/Zscaler.py]: done

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 5]
```diff
@@ -103,18 +104,18 @@ def http_request(method, url_suffix, data=None, headers=None, num_of_seconds_to_
             elif res.status_code in (401, 403):
                 raise AuthorizationError(res.content)
             elif res.status_code == 400 and method == 'PUT' and '/urlCategories/' in url_suffix:
-                raise Exception('Bad request, This could be due to reaching your organizations quota.'
-                                ' For more info about your quota usage, run the command zscaler-url-quota.')
+                return_error(
```
[the remark this answers is not in the document]
Comment (rootbid) [reply Packs/Zscaler/Integrations/Zscaler/Zscaler.py]: done

File: Packs/Zscaler/ReleaseNotes/1_3_10.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+##### Zscaler Internet Access
+
+Enhanced version of the content pack "Zscaler Internet Access". Changes:
+- Added command zscaler-list-ip-destination-groups
+- Added command zscaler-create-ip-destination-groups
+- Added command zscaler-edit-ip-destination-groups
+- Added command zscaler-list-ip-destination-groups-lite
+- Added command zscaler-delete-ip-destination-groups
+- Updated the Docker image to: *demisto/python3:3.10.11.57890*.
```
[the remark this answers is not in the document]
Comment (rootbid) [reply Packs/Zscaler/ReleaseNotes/1_3_10.md]: done

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Development container for Content. Poetry hash 1.11.0
+// Development container for Content. Poetry hash 022d65df3402ad71ca04cca0b76b759f0874f425330e511e952f4d2ebf9f265f
```
[the remark this answers is not in the document]
Comment (rootbid) [reply .devcontainer/devcontainer.json]: done

**Review by dantavori — APPROVED** [review]
great job!
I'm approving but please take a look at my note below. it's not required to change but is recommended.
let's run the commands and ensure they are working as expected before merging.

I will assign a reviewer to go over the documentation.

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 8]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      auto: PREDEFINED
+      predefined:
+      - "True"
+      - "False"
+      description: If set to true, the destination IP address group is non-editable. This field is applicable only to predefined IP address groups, which cannot be modified.
+      defaultValue: "False"
+      type: textArea
+    outputs:
+    - contextPath: Zscaler.IPDestinationGroup.ID
+      description: Unique identifier for the destination IP group
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1816]:
```suggestion
      description: Unique identifier for the destination IP group.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 9]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: number
+    - contextPath: Zscaler.IPDestinationGroup.Name
+      description: Destination IP group name
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1819]:
```suggestion
      description: Destination IP group name.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 10]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Type
+      description: Destination IP group type (i.e., the group can contain destination IP addresses or FQDNs)
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1822]:
```suggestion
      description: Destination IP group type (i.e., the group can contain destination IP addresses or FQDNs).
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 11]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Description
+      description: Destination IP group description
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1825]:
```suggestion
      description: Destination IP group description.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 12]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Addresses
+      description: Destination IP addresses, FQDNs, or wildcard FQDNs added to the group.
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.IpCategories
+      description: Destination IP address URL categories. You can identify destinations based on the URL category of the domain.
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Countries
+      description: Destination IP address counties. You can identify destinations based on the location of a server.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1834]:
```suggestion
      description: Destination IP address countries. You can identify destinations based on the location of a server.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 13]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    description: Updates the IP destination group information for the specified group ID.
+  - name: zscaler-list-ip-destination-groups
+    arguments:
+    - name: ip_group_id
+      description: a list of unique identifier for the IP destination groups.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1840]:
```suggestion
      description: A comma-separated list of unique identifiers for the IP destination groups.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 14]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      isArray: true
+      type: textArea
+    - name: exclude_type
+      auto: PREDEFINED
+      predefined:
+      - DSTN_IP
+      - DSTN_FQDN
+      - DSTN_DOMAIN
+      - DSTN_OTHER
+      description: The IP group type to be excluded from results.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1850]:
```suggestion
      description: The IP group type to be excluded from the results.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 15]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      - DSTN_FQDN
+      - DSTN_DOMAIN
+      - DSTN_OTHER
+      description: The IP group type to be filtered from results.
+    - name: include_ipv6
+      auto: PREDEFINED
+      predefined:
+      - "True"
+      - "False"
+      description: Retrieve IPv6 destination groups
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1866]:
```suggestion
      description: Retrieve IPv6 destination groups.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 16]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      defaultValue: "False"
+      type: textArea
+    - name: limit
+      description: Limit to results to be showcased, defaulted with the value 50 if nothing given as input
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1870]:
```suggestion
      description: Limit of the results to be retrieved.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 17]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      defaultValue: "50"
+      type: textArea
+    - name: all_results
+      auto: PREDEFINED
+      predefined:
+      - "True"
+      - "False"
+      description: Boolean value for user to choose, if wanted to get all results at once or not
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1878]:
```suggestion
      description: Whether to retrieve all results at once.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 18]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      defaultValue: "False"
+      type: textArea
+    - name: lite
+      auto: PREDEFINED
+      predefined:
+      - "True"
+      - "False"
+      description: Boolean value for user to choose, To retrieve only limited information of IP Destination groups. Includes ID, Name and Type of the IP Destination groups.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1886]:
```suggestion
      description: Whether to retrieve only limited information of IP destination groups. Includes ID, name and type of the IP destination groups.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 19]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      defaultValue: "False"
+      type: textArea
+    outputs:
+    - contextPath: Zscaler.IPDestinationGroup.ID
+      description: Unique identifier for the destination IP group
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1891]:
```suggestion
      description: Unique identifier for the destination IP group.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 20]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Name
+      description: Destination IP group name
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1894]:
```suggestion
      description: Destination IP group name.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 21]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Type
+      description: Destination IP group type (i.e., the group can contain destination IP addresses or FQDNs)
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1897]:
```suggestion
      description: Destination IP group type (i.e., the group can contain destination IP addresses or FQDNs).
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 22]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      description: Destination IP addresses, FQDNs, or wildcard FQDNs added to the group.
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Description
+      description: Additional information about the destination IP group
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.IpCategories
+      description: Destination IP address URL categories. You can identify destinations based on the URL category of the domain.
+      type: string
+    - contextPath: Zscaler.IPDestinationGroup.Countries
+      description: Destination IP address counties. You can identify destinations based on the location of a server.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1909]:
```suggestion
      description: Destination IP address countries. You can identify destinations based on the location of a server.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 23]
```diff
@@ -492,10 +492,1434 @@ script:
...
+      type: string
+    description: Gets a list of all IP destination groups or the IP destination group information for the specified group ID
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1911]:
```suggestion
    description: Gets a list of all IP destination groups or the IP destination group information for the specified group ID.
```

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.yml [hunk 24]
```diff
@@ -492,10 +492,1434 @@ script:
...
+  - name: zscaler-delete-ip-destination-groups
+    arguments:
+    - name: ip_group_ids
+      description: A list of unique identifier for the IP destination groups.
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/Integrations/Zscaler/Zscaler.yml:1915]:
```suggestion
      description: A comma-separated list of unique identifiers for the IP destination groups.
```

File: Packs/Zscaler/ReleaseNotes/1_3_10.md [hunk 25]
```diff
@@ -0,0 +1,10 @@
+
+#### Integrations
+
+##### Zscaler Internet Access
+- Add the following new commands:
```
Comment (ShirleyDenkberg) [inline Packs/Zscaler/ReleaseNotes/1_3_10.md:5]:
```suggestion
- Added the following new commands:
```

**Comment by ShirleyDenkberg** [comment]
@rootbid Please regenerate the README after implementing my comments in the yml file.
@dantavori Doc review completed.

File: Packs/Zscaler/Integrations/Zscaler/Zscaler.py [hunk 26]
```diff
@@ -1016,13 +1017,238 @@ def set_user_command(args):
...
+            cmd_url += exclude_type_param + type_params_str
+            responses = http_request('GET', cmd_url).json()
+            contents_filter = get_contents_lite(responses) if lite else get_contents(responses)
+            contents = contents_filter if all_results else contents_filter[:limit]
+            markdown = tableToMarkdown(
+                f"IPv4 Destination groups ({len(contents)})",
+                contents,
+                headers,
+                removeNull=True
+            )
```
[the remark this answers is not in the document]
Comment (rootbid) [reply Packs/Zscaler/Integrations/Zscaler/Zscaler.py]: we have removed code duplication from control statements, thanks!

**Comment by rootbid** [comment]
@ShirleyDenkberg 
README has been regenerated after implementing suggested changes in the yml file.
Thanks!

@dantavori 
We have tested all the commands once again in the XSOAR playground, they all work as expected.
The unit tests have been fixed as well.

**Comment by rootbid** [comment]
> @rootbid seems like there are still some mypy errors, also need to update the docker image. can you handle please? last thing before merging.

- mypy errors have been fixed
- docker image updated to `demisto/python3:3.10.12.62631`

File: Packs/Zscaler/ReleaseNotes/1_3_10.md [hunk 27]
```diff
@@ -0,0 +1,10 @@
+
+#### Integrations
+
+##### Zscaler Internet Access
+- Added the following new commands:
+  - ***zscaler-list-ip-destination-groups***
+  - ***zscaler-create-ip-destination-groups***
+  - ***zscaler-edit-ip-destination-groups***
+  - ***zscaler-delete-ip-destination-groups***
+- Updated the Docker image to: *demisto/python3:3.10.11.61265*.
```
Comment (dantavori) [inline Packs/Zscaler/ReleaseNotes/1_3_10.md:10]:
```suggestion
- Updated the Docker image to: * demisto/python3:3.10.12.62631*.
```

**Pull request merged**